### PR TITLE
Add more tests for pod preemption

### PR DESCRIPTION
**Pull request by bsalamat**
**What this PR does / why we need it**:
Adds more e2e and integration tests for pod preemption.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
This PR is based on #50949. Only the last commit is new.

**Release note**:

```release-note
NONE
```

ref/ #47604

@kubernetes/sig-scheduling-pr-reviews @davidopp

### CODE REVIEW

File: test/e2e/scheduling/preemption.go [hunk 1]
```diff
@@ -125,4 +125,125 @@ var _ = SIGDescribe("SchedulerPreemption [Serial] [Feature:PodPreemption]", func
 			Expect(livePod.DeletionTimestamp).To(BeNil())
 		}
 	})
+
+	// This test verifies that when a high priority pod is pending and its
+	// scheduling violates a medium priority pod anti-affinity, the medium priority
+	// pod is preempted to allow the higher priority pod schedule.
+	// It also verifies that existing low priority pods are not preempted as their
+	// preemption wouldn't help.
+	It("validates pod anti-affinity works in preemption", func() {
```
Comment (timothysc) [inline test/e2e/scheduling/preemption.go]: Isn't this a pretty expensive test if you have a large cluster, do we want it defaulted on or behind [Feature] ?

**Review by timothysc — CHANGES_REQUESTED** [review]
So I don't really have an issue with the basics of the test, but we need to be mindful that it will be used on the scale tests and in other labs.  So this will schedule 5k+ pods on the scale test 

/cc @wojtek-t

File: test/e2e/scheduling/preemption.go (hunk 1 above)
Comment (bsalamat) [reply test/e2e/scheduling/preemption.go]: It is behind [Feature] right now. How can we prevent it from being used in scale tests?

Comment (timothysc) [reply test/e2e/scheduling/preemption.go]: Ahh I scrolled back and see it now... 

Is there any reason you need to hit * nodes vs a subset?  It's still really expensive.

Comment (bsalamat) [reply test/e2e/scheduling/preemption.go]: Yes, there is a reason. In order to trigger preemption, cluster should be under resource pressure.

Comment (timothysc) [reply test/e2e/scheduling/preemption.go]: Can't you use node selectors to force it?

Comment (bsalamat) [reply test/e2e/scheduling/preemption.go]: Yes, I could, but the current scenario is better, as it also exercises node scoring and ensures that we preempt the right pod no matter where in the cluster it is scheduled and also makes sure that other pods are not impacted. Node selector would remove these benefits.
What I heard is that the CI clusters typically have 4-5 nodes, except those that run scalability tests. If this is true, then the test shouldn't be very time consuming.

Comment (timothysc) [reply test/e2e/scheduling/preemption.go]: What I'm really concerned about is the e2e tests are used in the field to validate clusters.  There are times when they want to validate that features are working properly.  If I was company (X) trying to validate a 200 node cluster you'd be loading their cluster and this is a very expensive test.  The usage of the e2e's goes well beyond upstream CI.

Comment (timothysc) [reply test/e2e/scheduling/preemption.go]: Perhaps using a [Disruptive] tag on this test.

Comment (bsalamat) [reply test/e2e/scheduling/preemption.go]: I see your point. I just changed the test to create a maximum of 4 pods.

Comment (timothysc) [reply test/e2e/scheduling/preemption.go]: thx for the change.

File: test/e2e/scheduling/preemption.go [hunk 2]
```diff
@@ -125,4 +125,130 @@ var _ = SIGDescribe("SchedulerPreemption [Serial] [Feature:PodPreemption]", func
...
+							},
+						},
+					},
+				},
+			},
+		})
+		// Make sure that the medium priority pod on the first node is preempted to
+		// allow the high priority pod run.
+		preemptedPod, err := cs.CoreV1().Pods(pods[0].Namespace).Get(pods[0].Name, metav1.GetOptions{})
+		podDeleted := (err != nil && errors.IsNotFound(err)) ||
```
Comment (timothysc) [inline test/e2e/scheduling/preemption.go]: This logic is a little conflating... more a comment the please change.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve no-issue

File: test/e2e/scheduling/preemption.go (hunk 2 above)
Comment (bsalamat) [reply test/e2e/scheduling/preemption.go]: Sorry, I didn't get the problem. Do you want me to remove the last part of the comment: "to allow the high priority pod run"?

**Pull request merged**